fdbclient/TagThrottle.h

```diff
@@ -55,55 +55,71 @@ class TagSet {
 		return tags.end();
 	}
 
-//private:
-	Arena arena;
-	std::set<TransactionTagRef> tags;
-	size_t bytes;
-};
-
-template <>
-struct dynamic_size_traits<TagSet> : std::true_type {
-	// May be called multiple times during one serialization
 	template <class Context>
-	static size_t size(const TagSet& t, Context&) {
-		return t.tags.size() + t.bytes;
-	}
-
-	// Guaranteed to be called only once during serialization
-	template <class Context>
-	static void save(uint8_t* out, const TagSet& t, Context& c) {
+	void save(uint8_t* out, Context& c) const {
 		uint8_t *start = out;
-		for (const auto& tag : t.tags) {
+		for (const auto& tag : *this) {
 			*(out++) = (uint8_t)tag.size();
 
 			std::copy(tag.begin(), tag.end(), out);
 			out += tag.size();
 		}
 
-		ASSERT((size_t)(out-start) == size(t, c));
+		ASSERT((size_t)(out - start) == size() + bytes);
 	}
 
-	// Context is an arbitrary type that is plumbed by reference throughout the
-	// load call tree.
 	template <class Context>
-	static void load(const uint8_t* data, size_t size, TagSet& t, Context& context) {
+	void load(const uint8_t* data, size_t size, Context& context) {
 		//const uint8_t *start = data;
 		const uint8_t *end = data + size;
 		while(data < end) {
 			uint8_t len = *(data++);
-			TransactionTagRef tag(context.tryReadZeroCopy(data, len), len);
 			data += len;
-
-			t.tags.insert(tag);
-			t.bytes += tag.size();
+			// Tags are already deduplicated
+			const auto& tag = tags.emplace_back(context.tryReadZeroCopy(data, len), len);
```

Comment:
I notice that `data += len` has been reordered with respect to when we read from `data`. Is that intentional?